feat: add delete option to three dots menu

johnpc opened the pull request:
<!--
  This template provides some ideas of things to include in your PR description.

  To start, try providing a short summary of your changes in the Title above. We follow Conventional Commits syntax, please ensure your title is prefixed with one of:
  - `feat: `
  - `fix: `
  - `docs: `
  - `chore: `
  - `dev:`

  If a section of the PR template does not apply to this PR, then delete that section.

  PLEASE READ:
  -------------------------
  Mealie is moving to a regular, automatic release schedule. This means that all PRs should be in a
  stable state, ready for release. This includes:

  - Ensuring new tests have been added to cover new features, or to prevent regressions.
  - Work is fully complete and usable

 -->

## What type of PR is this?

- feature

## What this PR does / why we need it:

This PR adds the "Delete" option to the three dots menu. Excluding delete from the three dots menu has caused confusion as seen in https://github.com/mealie-recipes/mealie/discussions/3984#discussioncomment-11748286

Before: 

![Screenshot 2025-01-06 at 12 40 27 PM](https://github.com/user-attachments/assets/31601ba9-8776-4e51-b79c-39642f9a5420)

After: 

![Screenshot 2025-01-06 at 12 33 26 PM](https://github.com/user-attachments/assets/19e55f98-fe6f-4332-978b-4d1e789e59d9)

## Which issue(s) this PR fixes:

No issue, but feature request is here: https://github.com/mealie-recipes/mealie/discussions/3984#discussioncomment-11748286

## Special notes for your reviewer:

This may have been omitted intentionally from the three dots menu, considering all the plumbing was already in place and I just had to flip the value to `true`. Was this omitted for a good reason to where we actually don't want to allow this?

## Testing

I tested manually, but I'm not sure where test coverage should be added. Please advise.

johnpc committed:
feat: add delete option to three dots menu

simgunz commented:
Great to see this feature implemented. A minor suggestion: I believe that the best practice in terms of UX is to place the 'Delete' action as the last item in the menu.

Kuchenpirat reviewed (commented):
While i agree with  the general sentiment that the delete option might be a bit too hidden i don't think we need to include it on the RecipeExplore page. 
The implementation on this page also comes with the problem that the state of the recipe page does not reflect that the recipe has been deleted. 

Where i can see the delete button making sense is in the RecipeActionMenu, no longer hiding it behind the edit button. 
As @simgunz suggested i would move it into last position (above the divider) and i would keep it grey instead of red (so that it isnt the first thing your eyes get drawn towards when opening the menu).
Here we also need add functionality to navigate away from the page and reload the recipes as the current delete button does as well.

johnpc commented:
@Kuchenpirat I've made the update as you've suggested. 

* Only added the delete action to the three dots in the detail view (not list view)
* Added handling so that that the delete action navigates back to the list view once deleted
* Changed the delete color from red to gray

<img width="265" alt="Screenshot 2025-01-14 at 7 22 24 PM" src="https://github.com/user-attachments/assets/1acc8520-6438-4912-a0d0-57ea5ea9d58f" />

ChipSkylark37 commented:
Does this still prompt the normal "Are you sure" confirmation on this? 

With it being close to the 'Share' button, I know I would accidentally do this _at least_ once.

michael-genson commented:
Might also be worth putting a divider in-between share and delete to visually separate it (similar to data management)

![image](https://github.com/user-attachments/assets/ce9c7f58-939d-4a1a-b128-382f8c55a9b3)

johnpc commented:
@ChipSkylark37  Yes, the confirmation prompt is still there. Here is a screen recording


https://github.com/user-attachments/assets/850d747e-ed82-42c6-9088-d7ddceba95a4

johnpc commented:
@michael-genson I'm not a designer but in my mind, the sections in the example you shared work well because those dividers actually divide logical sections. In this example it would just be to create space away from the delete action, which doesn't feel like the same thing. I think the confirmation dialog is enough to prevent disaster in the case of accidental clicks.

johnpc committed:
Merge branch 'mealie-next' into add-delete-threedots

michael-genson commented:
Yup, that makes sense! Adding a divider probably doesn't do much

I was thinking of some apps where "dangerous" operations are visually separate from everything else, but other than delete we don't really have anything like that

simgunz commented:
> While i agree with the general sentiment that the delete option might be a bit too hidden i don't think we need to include it on the RecipeExplore page.

Is the RecipeExplore page the one that lists all the recipes (the one in the picture below)?

![Screenshot_20250115_082027](https://github.com/user-attachments/assets/a261c181-2f37-4cbf-905b-e78f7f98dcfb)

I believe that the delete button should _also_ be present in that three-dots menu. From my user experience, it is a common UX pattern that of being able to delete an item from a list view (see the Gmail UI for example where there are quick actions to delete an email).

![Screenshot_20250115_082409](https://github.com/user-attachments/assets/a6b4ba55-5e38-4f90-a1b1-3cf6a6a68b19)

Kuchenpirat commented:
I don't think this is an equal situation as in my experience I am way more likely to delete a mail than a recipe I or someone in my household added with some intend.
But I'll think about it 👍

simgunz commented:
You convinced me. Deleting a recipe is indeed a seldom action. Adding the
action in the detail view menu is already good enough.

On Wed, Jan 15, 2025, 08:55 Kuchenpirat ***@***.***> wrote:

> I don't think this is an equal situation as in my experience I am way more
> likely to delete a mail than a recipe I or someone in my household added
> with some intend.
> But I'll think about it 👍
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/mealie-recipes/mealie/pull/4842#issuecomment-2591866229>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADR2XV7XZCHGS455A3F2CL2KYH5LAVCNFSM6AAAAABUWB54EOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDKOJRHA3DMMRSHE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

johnpc commented:
I think it's fine to only have it on the detail page personally.

When I was first trying to delete a recipe, the recipe detail page three dots menu (the one in this PR in current form) was the first place I checked. When it wasn't there I also checked the list view page three dots menu. 

When it wasn't there I got very confused, poked around a while longer to no avail, and eventually googled it and found https://github.com/mealie-recipes/mealie/discussions/3984#discussioncomment-11748286

Then a few weeks later, I ran into the _same situation again_! I'd forgotten how to delete a recipe, and had to google it again to re-find that discussion thread explaining how! That's when I was like okay I'll just cut a PR.

ChipSkylark37 commented:
> I think it's fine to only have it on the detail page personally.
> 
> When I was first trying to delete a recipe, the recipe detail page three dots menu (the one in this PR in current form) was the first place I checked. When it wasn't there I also checked the list view page three dots menu.
> 
> When it wasn't there I got very confused, poked around a while longer to no avail, and eventually googled it and found [#3984 (comment)](https://github.com/mealie-recipes/mealie/discussions/3984#discussioncomment-11748286)
> 
> Then a few weeks later, I ran into the _same situation again_! I'd forgotten how to delete a recipe, and had to google it again to re-find that discussion thread explaining how! That's when I was like okay I'll just cut a PR.

I like the option into the three dots menu (without going into 'Edit').

I also agree it should only be on the Individual Recipe Page itself, not from the page that shows all Recipes.

Kuchenpirat reviewed (approved):
Thanks for helping out with the project 😊👍

LGTM 🚀

Kuchenpirat committed:
Merge branch 'mealie-next' into add-delete-threedots
